File: 문제풀이/온라인 저지/프로그래머스/# 02. Level 2/# 042863. 큰 수 만들기/kotlin/Solution.kt
class Solution {
    fun solution(number: String, k: Int): String {
        var k = k
        val dq = Deque(number.length)
        for (n in number) {
            while (dq.isNotEmpty() && dq.last() < n && k > 0) {
                dq.dqLast()
                k--
            }
            dq.eq(n)
        }
        while (k > 0) {
            dq.dqLast()
            k--
        }
        val sb = StringBuilder()
        while (dq.isNotEmpty()) {
            sb.append(dq.dqFirst())
        }
        return sb.toString()
    }
}

class Deque(cap: Int) {
    private val c = cap + 1
    private val arr = CharArray(c)
    private var front = 0
    private var rear = 0

    fun eq(e: Char) {
        arr[((rear + 1 + c) % c).also { rear = it }] = e
    }

    fun dqFirst(): Char = arr[((front + 1 + c) % c).also { front = it }]
    fun dqLast(): Char = arr[rear.also { rear = (it - 1 + c) % c }]
    fun last(): Char = arr[rear]
    fun isNotEmpty(): Boolean = front != rear
}

File: 문제풀이/온라인 저지/프로그래머스/# 02. Level 2/# 042863. 큰 수 만들기/python/solution.py
def solution(number, k):
    stack = []
    for num in number:
        # 앞에서 작은 수가 있고, 삭제횟수가 있으면 가능한 지워야함
        while stack and k > 0 and num > stack[-1]:
            stack.pop()
            k -= 1
        stack.append(num)

    # 삭제횟수를 다 못 쓸 수 있음. 이 경우 뒤에 쌓이는 숫자들은 점점 갈 수록 숫자가 작아짐
    # 뒤의 숫자들을 차례로 제거해야함
    while k > 0:
        stack.pop()
        k -= 1
    return ''.join(stack)
